src/components/implementation/tests/unit_heap/heap_user.c

```diff
@@ -0,0 +1,155 @@
+    return ((hentry_t *)a)->value <= ((hentry_t *)b)->value;
+}
+int
+c_max(void *a, void *b)
+{
+    return ((hentry_t *)a)->value >= ((hentry_t *)b)->value;
+}
+void
+u(void *e, int pos)
+{
+    ((hentry_t *)e)->index = pos;
+}
+
+DECLARE_HEAP(min, c_min, u);
+DECLARE_HEAP(max, c_max, u);
+
+static void
+entries_validate(heap_t *h, hentry_t *es, int amnt)
+{
+    int i;
+
+    for (i = 0; i < amnt; i++) {
+        assert(h->data[es[i].index] == &es[i]);
+    }
+}
+
+static void
+test_driver(int amnt, int type)
+{
+    int            i;
+    hentry_t *prev, *es;
+    heap_t *  h;
+
+    h  = heap_alloc(amnt);
+    es = malloc(sizeof(hentry_t) * amnt);
+    assert(es);
+
+    for (i = 0; i < amnt; i++) {
+        es[i].value = rand() % VAL_BOUND;
+        printc("adding@%d:%d\n", i, es[i].value);
+        if (type == MIN) {
+            assert(!min_heap_add(h, &es[i]));
+            assert(!min_heap_verify(h, 1));
+        } else {
+            assert(!max_heap_add(h, &es[i]));
+            assert(!max_heap_verify(h, 1));
+        }
+    }
+    entries_validate(h, es, amnt);
+    for (i = 0; i < amnt; i++) {
+        es[i].value = rand() % VAL_BOUND;
+        printc("adjusting@%d:%d\n", i, es[i].value);
+        if (type == MIN) {
+            min_heap_adjust(h, es[i].index);
+            assert(!min_heap_verify(h, 1));
+        } else {
+            max_heap_adjust(h, es[i].index);    
+            assert(!max_heap_verify(h, 1));
+        }
+    }
+    entries_validate(h, es, amnt);
+    prev = h->data[1];
+    for (i = 0; i < amnt; i++) {
+        hentry_t *curr = type == MIN ? min_heap_highest(h) : max_heap_highest(h);
+        printc("highest:%d\n", curr->value);
+        if (type == MIN) {    
+            if (!c_min((hentry_t *)prev, (hentry_t *)curr)) assert(0);
+        } else {
+            if (!c_max((hentry_t *)prev, (hentry_t *)curr)) assert(0);
+        }
+        prev = curr;
+    }
+    if (type == MIN) {
+        assert(!min_heap_highest(h));
+        assert(!min_heap_verify(h, 1));
+    } else {
+        assert(!max_heap_highest(h));
+        assert(!max_heap_verify(h, 1));
+    }
+    
+    assert(heap_size(h) == 0);
+    for (i = 0; i < amnt; i++) {
+        es[i].value = rand() % VAL_BOUND;
+        printc("adding@%d:%d\n", i, es[i].value);
+        assert(!min_heap_add(h, &es[i]));
+    }
+    entries_validate(h, es, amnt);
+    for (i = amnt; i > 0; i--) {
+        int idx;
+        idx                 = (rand() % i) + 1;
+        hentry_t *curr = 
+        type == MIN ? min_heap_remove(h, idx) : max_heap_remove(h, idx);
+        printc("removing:%d\n", curr->value);
+        assert(curr);
+        assert(h->e == i);
+    }
+    
+    if (type == MIN) {
+        assert(!min_heap_highest(h));
+        assert(!min_heap_verify(h, 1));
+    } else {
+        assert(!max_heap_highest(h));
+        assert(!max_heap_verify(h, 1));
+    } 
+    assert(heap_size(h) == 0);
+
+    heap_destroy(h);
+    free(es);
+}
+
+#define ITER 10
```

Comment:
The rest of the code uses tabs. This uses spaces.

Please use tabs. I believe we have an autoformat script lying around that you might apply to this to fix it. Don't remember where, sorry!!!!

Reply:
Oh, I think copy paste lead to this. Thanks, fixed now.